### PR TITLE
fix flaky-test: AuthorizationTest.simple

### LINKED ISSUE #9709: Flaky-test: AuthorizationTest.simple
AuthorizationTest is flaky. The simple test method fails sporadically.

The assertion error shown in logs is misleading. The assertion error happens as a result of a retry. The actual problem can be found by downloading the surefire artifacts file. This is the NPE that happens:

```
java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[?:1.8.0_252]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928) ~[?:1.8.0_252]
        at org.apache.pulsar.broker.authorization.AuthorizationService.canLookup(AuthorizationService.java:269) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
        at org.apache.pulsar.broker.web.PulsarWebResource.checkAuthorization(PulsarWebResource.java:792) ~[classes/:?]
        at org.apache.pulsar.broker.web.PulsarWebResource.checkConnect(PulsarWebResource.java:782) ~[classes/:?]
        at org.apache.pulsar.broker.admin.AdminResource.getPartitionedTopicMetadata(AdminResource.java:539) ~[classes/:?]
        at org.apache.pulsar.broker.admin.impl.PersistentTopicsBase.internalGrantPermissionsOnTopic(PersistentTopicsBase.java:348) ~[classes/:?]
        at org.apache.pulsar.broker.admin.v1.PersistentTopics.grantPermissionsOnTopic(PersistentTopics.java:126) ~[classes/:?]
...
Caused by: java.lang.NullPointerException
	at java.util.TreeMap.getEntry(TreeMap.java:347) ~[?:1.8.0_252]
	at java.util.TreeMap.get(TreeMap.java:278) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$checkPermission$12(PulsarAuthorizationProvider.java:404) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.checkPermission(PulsarAuthorizationProvider.java:397) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.checkAuthorization(PulsarAuthorizationProvider.java:378) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$canConsumeAsync$2(PulsarAuthorizationProvider.java:148) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.canConsumeAsync(PulsarAuthorizationProvider.java:112) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.authorization.AuthorizationService.lambda$canConsumeAsync$1(AuthorizationService.java:216) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456) ~[?:1.8.0_252]
	... 1 more
```

<details>
<summary>Full log from artifacts/artifacts/pulsar-broker/target/surefire-reports/org.apache.pulsar.broker.auth.AuthorizationTest-output.txt file</summary>
<code><pre>
------- Starting test [TestClass name=class org.apache.pulsar.broker.auth.AuthorizationTest].simple([])-------
22:36:21.683 [pulsar-web-5980-7] INFO  org.apache.pulsar.broker.admin.impl.ClustersBase - [null] Created cluster c1
22:36:21.684 [pulsar-web-5980-7] INFO  org.eclipse.jetty.server.RequestLog - 127.0.0.1 - - [24/Feb/2021:22:36:21 +0000] "PUT /admin/v2/clusters/c1 HTTP/1.1" 204 0 "-" "Pulsar-Java-v2.8.0-SNAPSHOT" 4
22:36:21.697 [metadata-store-5956-1] INFO  org.apache.pulsar.broker.admin.impl.TenantsBase - [null] Created tenant p1
22:36:21.698 [metadata-store-5956-1] INFO  org.eclipse.jetty.server.RequestLog - 127.0.0.1 - - [24/Feb/2021:22:36:21 +0000] "PUT /admin/v2/tenants/p1 HTTP/1.1" 204 0 "-" "Pulsar-Java-v2.8.0-SNAPSHOT" 4
22:36:21.804 [pulsar-web-5980-6] INFO  org.apache.pulsar.broker.web.PulsarWebResource - Successfully validated clusters on tenant [p1]
22:36:21.806 [pulsar-web-5980-6] INFO  org.apache.pulsar.broker.admin.impl.NamespacesBase - [null] Created namespace p1/c1/ns1
22:36:21.807 [pulsar-web-5980-6] INFO  org.eclipse.jetty.server.RequestLog - 127.0.0.1 - - [24/Feb/2021:22:36:21 +0000] "PUT /admin/namespaces/p1/c1/ns1 HTTP/1.1" 204 0 "-" "Pulsar-Java-v2.8.0-SNAPSHOT" 7
22:36:21.958 [pulsar-web-5980-8] INFO  org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider - [my-role] Successfully granted access for role my-role: [produce] - namespace p1/c1/ns1
22:36:21.958 [pulsar-web-5980-8] INFO  org.eclipse.jetty.server.RequestLog - 127.0.0.1 - - [24/Feb/2021:22:36:21 +0000] "POST /admin/namespaces/p1/c1/ns1/permissions/my-role HTTP/1.1" 204 0 "-" "Pulsar-Java-v2.8.0-SNAPSHOT" 4
22:36:22.078 [Thread-451] WARN  org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider - Client  with Role - null failed to get permissions for topic - persistent://p1/c1/ns1/ds2. java.lang.NullPointerException
22:36:22.087 [Thread-452] WARN  org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider - Client  with Role - null failed to get permissions for topic - persistent://p1/c1/ns1/ds2. java.lang.NullPointerException
22:36:22.087 [Thread-452] WARN  org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider - Client with Role - null failed to get permissions for topic - persistent://p1/c1/ns1/ds2. java.lang.NullPointerException
22:36:22.087 [pulsar-web-5980-5] WARN  org.apache.pulsar.broker.authorization.AuthorizationService - Role - null failed to get lookup permissions for topic - persistent://p1/c1/ns1/ds2. java.lang.NullPointerException
22:36:22.087 [pulsar-web-5980-5] WARN  org.apache.pulsar.broker.admin.AdminResource - Unexpected error while authorizing lookup. topic=persistent://p1/c1/ns1/ds2, role=null. Error: java.lang.NullPointerException
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.AuthorizationService.canLookup(AuthorizationService.java:269) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.web.PulsarWebResource.checkAuthorization(PulsarWebResource.java:792) ~[classes/:?]
	at org.apache.pulsar.broker.web.PulsarWebResource.checkConnect(PulsarWebResource.java:782) ~[classes/:?]
	at org.apache.pulsar.broker.admin.AdminResource.getPartitionedTopicMetadata(AdminResource.java:539) ~[classes/:?]
	at org.apache.pulsar.broker.admin.impl.PersistentTopicsBase.internalGrantPermissionsOnTopic(PersistentTopicsBase.java:348) ~[classes/:?]
	at org.apache.pulsar.broker.admin.v1.PersistentTopics.grantPermissionsOnTopic(PersistentTopics.java:126) ~[classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_252]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_252]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$VoidOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:159) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791) ~[jetty-servlet-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1626) ~[jetty-servlet-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.apache.pulsar.broker.web.ResponseHandlerFilter.doFilter(ResponseHandlerFilter.java:58) ~[classes/:?]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193) ~[jetty-servlet-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601) ~[jetty-servlet-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548) ~[jetty-servlet-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1435) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501) ~[jetty-servlet-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1350) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:234) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:179) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.Server.handle(Server.java:516) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388) ~[jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633) [jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380) [jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273) [jetty-server-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [jetty-io-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [jetty-io-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) [jetty-io-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [jetty-util-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [jetty-util-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [jetty-util-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [jetty-util-9.4.35.v20201120.jar:9.4.35.v20201120]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375) [jetty-util-9.4.35.v20201120.jar:9.4.35.v20201120]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.51.Final.jar:4.1.51.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
Caused by: java.lang.NullPointerException
	at java.util.TreeMap.getEntry(TreeMap.java:347) ~[?:1.8.0_252]
	at java.util.TreeMap.get(TreeMap.java:278) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$checkPermission$12(PulsarAuthorizationProvider.java:404) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.checkPermission(PulsarAuthorizationProvider.java:397) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.checkAuthorization(PulsarAuthorizationProvider.java:378) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$canConsumeAsync$2(PulsarAuthorizationProvider.java:148) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.canConsumeAsync(PulsarAuthorizationProvider.java:112) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at org.apache.pulsar.broker.authorization.AuthorizationService.lambda$canConsumeAsync$1(AuthorizationService.java:216) ~[pulsar-broker-common-2.8.0-SNAPSHOT.jar:2.8.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456) ~[?:1.8.0_252]
	... 1 more
</pre></code>
</details>

### PR DESCRIPTION
**Pull request by 315157973**
Fixes #9709

### Motivation
```
Caused by: java.lang.NullPointerException
	at java.util.TreeMap.getEntry(TreeMap.java:347)
```
From the stack trace, it is caused by the key read by `TreeMap` being null. Therefore, the role must be null.
```
//PulsarAuthorizationProvider.java:404
Map<String, Set<AuthAction>> namespaceRoles = policies.get().auth_policies.namespace_auth;
Set<AuthAction> namespaceActions = namespaceRoles.get(role);
```

According to the code, the role is read through `httpRequest.getAttribute` in `PulsarWebResource#clientAppId`, so it is normal to be null.

Although the role is null, why do unit tests pass occasionally?
I added logs and found that the implementation class of Map is not always TreeMap, but occasionally HashMap.
When it is a HashMap, no error will be reported if the key is null. This problem is caused by Jackson's serialization.

I checked AuthPolicies, all Maps are initialized as TreeMap, but we have never used any features of TreeMap. 
In order to avoid serialization problems elsewhere, I changed TreeMap to HashMap.
Another way is to add the annotation @JsonDeserialize(as = TreeMap.class), and then make the judgment that the key is null.
Why was TreeMap used here? May need @merlimat  to help explain.

### CODE REVIEW

**Comment by lhotari** [comment]
@315157973 This won't fix #9711 . I have put the explanation about the thread safety issue as a comment in https://github.com/apache/pulsar/issues/9711#issuecomment-786631166 .

**Comment by lhotari** [comment]
The issue in #9711 is most likely different than the issue which makes `AuthorizationTest.simple` flaky.

**Comment by lhotari** [comment]
I created an alternative fix which fixes a problem with the test itself: #9746 . I put some notes in the description about the relation to this fix. It would be interesting to find out why the test was flaky and didn't fail always. That is still a mystery to reveal.

**Comment by 315157973** [comment]
> I created an alternative fix which fixes a problem with the test itself: #9746 . I put some notes in the description about the relation to this fix.
> 
> > Although the role is null, why do unit tests pass occasionally?
> > I added logs and found that the implementation class of Map is not always TreeMap, but occasionally HashMap.
> > When it is a HashMap, no error will be reported if the key is null. This problem is caused by Jackson's serialization.
> 
> I guess this explains why it passed in some cases. It's odd that the behavior is so nondeterministic. Why does the serialization get used at all? I had the impression that there's a local cache that will be used if the entries are available locally, however I haven't checked that aspect.
> 
> btw. The null value for the role shouldn't be a normal case. The fix in #9746 fixes the test itself that was invalid.

As long as it is serialized and deserialized once, it will become a hashMap. Because zk update takes time, sometimes it is still TreeMap locally.

**Comment by eolivelli** [comment]
I am fine with having the HashMap and use it consistently.
so +1 from me to this patch.
 
This is definitely a good idea.
Using TreeMap consistently is okay as well, but if usually we have an HashMap (in the majority of cases) it is better to go with HashMap.
It would be better to have some way to ensure that we won't change it again in the future with some refactor.
So I don't see this patch as a direct fix for the test.

by the way I share @lhotari concerns, we should add explicit checks about security/auth data structures and consistency of the runtime environment.


so probably the best is to have both of the two patches merged.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -323,7 +323,7 @@ private void grantPermissions(String topicUri, String role, Set<AuthAction> acti
         try {
             namespaceResources().set(path(POLICIES, namespaceName.toString()), (policies) -> {
                 if (!policies.auth_policies.destination_auth.containsKey(topicUri)) {
-                    policies.auth_policies.destination_auth.put(topicUri, new TreeMap<String, Set<AuthAction>>());
+                    policies.auth_policies.destination_auth.put(topicUri, new HashMap<>());
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:326]: why don't we use Maps.newHashMap ?

**Review by eolivelli — APPROVED** [review]

**Review by aahmed-se — APPROVED** [review]

**Pull request merged**